File: Google/Golang/binary-tree-maximum-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
    res := ^int(^uint32(0) >> 1)
    helper(root, &res)
    return res
}

func helper(root *TreeNode, res *int) int {
    if root == nil {
        return 0
    }
    left := helper(root.Left, res)
    right := helper(root.Right, res)
    pathSum := max(max(root.Val,left+root.Val+right), max(left+root.Val, right+root.Val))
    if pathSum > *res {
        *res = pathSum
    }
    return max(root.Val, max(left+root.Val, right+root.Val))
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: Tree/C++/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int res = INT_MIN;
        if (root == NULL) {
            return 0;
        }
        helper(root, res);
        return res;
    }
    
    int helper(TreeNode *root, int &res) {
        int left, right, tmp;
        if (root == NULL) {
            return 0;
        }
        left = helper(root->left, res);
        right = helper(root->right, res);
        tmp = max(
            max(root->val, left+root->val+right),
            max(left+root->val, right+root->val)
        );
        res = max(res, tmp);
        return max(root->val, max(left+root->val, right+root->val));
    }
};
